### DISCUSSION TITLE
Any tips/gotcha's for upgrading from version 2 (2.9.12) to version 4 (4.3.1) for Solar Only?

### DISCUSSION THREAD

**Discussion by hulkster** [jasonacox/Powerwall-Dashboard#475] [General]
So I have been lazy and haven't been upgrading Powerwall-Dashboard for a while ... but HEY, it just keeps working great!   ;-)

I'm currently running 2.9.12 and the latest version is 4.3.1 ... and FYI that I'm Solar Only - no Powerwall.

So is it simply a matter of running upgrade.sh ... and everything will work and I'll retain my historical energy and weather data?

I know there has been some merging of the Solar Only code, so figured worth a quick ask since I'm now two major versions back.

Thanks!

**Comment by jasonacox** [comment]
Hi @hulkster - We do what we can to make the upgrade process seamless by using `upgrade.py`.   Having said that, I recommend making sure you have a good backup before you upgrade, just in case.

If it helps, you could install a fresh install on another computer to test.

**Comment by hulkster** [comment]
So I tried using upgrade.py and ran into an issue. So then I tried starting completely from scratch and ran into (I think) the same issue.

I have Solar-Only ... so I select option number 2 - Tesla Cloud. 
I use a different uid/gid that 1000/1000, but setup.sh figures that out (recall I was involved in that change).

All seems to go well (also setup Weather411) with influxdb, grafana, pypowerwall, weather411, and telegraf all Running/Started ... and then I have to authenticate my Tesla Account which is a simple copy-paste and then copy-paste ... and then I get:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.10/site-packages/pypowerwall/__main__.py", line 80, in <module>
    if c.setup(email):
  File "/usr/local/lib/python3.10/site-packages/pypowerwall/cloud/pypowerwall_cloud.py", line 923, in setup
    if not self.connect():
  File "/usr/local/lib/python3.10/site-packages/pypowerwall/cloud/pypowerwall_cloud.py", line 181, in connect
    f"({sites[self.siteindex]['site_name']}) for {self.email}")
KeyError: 'site_name'


$ ./verify.sh
Verify Powerwall-Dashboard 4.3.1 on Linux - Timezone: America/Denver

This script will attempt to verify all the services needed to run
Powerwall-Dashboard. Use this output when you open an issue for help:
https://github.com/jasonacox/Powerwall-Dashboard/issues/new


Checking pypowerwall
----------------------------------------------------------------------------
 - Config File pypowerwall.env: GOOD
 - Container (pypowerwall): GOOD
 - Service (port 8675): ERROR: Not Listening
 - Version: Unknown
 - Powerwall State: ERROR: Not Connected - Firmware Version: Unknown
 - Cloud Mode: YES - ERROR: Not Connected to Tesla Cloud

Checking telegraf
----------------------------------------------------------------------------
 - Config File telegraf.conf: GOOD
 - Local Config File telegraf.local: GOOD
 - Container (telegraf): GOOD
 - Version: Telegraf 1.28.2 (git: HEAD@8d9cf395)

Checking influxdb
----------------------------------------------------------------------------
 - Config File influxdb.conf: GOOD
 - Environment File influxdb.env: GOOD
 - Container (influxdb): GOOD
 - Service (port 8086): GOOD
 - Filesystem (./influxdb): GOOD
 - Version: InfluxDB shell version: 1.8.10

Checking grafana
----------------------------------------------------------------------------
 - Config File grafana.env: GOOD
 - Container (grafana): GOOD
 - Service (port 9000): GOOD
 - Filesystem (./grafana): GOOD
 - Version: Grafana CLI version 9.1.2

Checking weather411
----------------------------------------------------------------------------
 - Container (weather411): GOOD
 - Service (port 8676): GOOD
 - Weather: {"temperature": 49.03}
 - Version: 0.2.3


$ docker logs -f pypowerwall
05/14/2024 06:56:37 PM [proxy] [INFO] pyPowerwall [0.8.4] Proxy Server [t55] - HTTP Port 8675
05/14/2024 06:56:37 PM [proxy] [INFO] pyPowerwall Proxy Started
05/14/2024 06:56:37 PM [pypowerwall.cloud.pypowerwall_cloud] [WARNING] Missing auth file .auth/.pypowerwall.auth - run setup
05/14/2024 06:56:37 PM [proxy] [ERROR] Missing auth file .auth/.pypowerwall.auth - run setup
05/14/2024 06:56:37 PM [proxy] [ERROR] Fatal Error: Unable to connect. Please fix config and restart.

Note that .auth/.pypowerwall.auth was (correctly) written  ... and if I re-run setup.sh it knows it is there ... but ... I still get the same error.

So I must be doing something really stupid here (although this should be pretty simple) and I'm not a Python guy ... so a bit hard for me to look into this deeper. 

Any suggestions on stuff to try? 

**BTW, I ALSO tried a completely clean install by removing docker from the system and re-installing, change the user:group of the Solar account to 1000:1000, pulled a clean copy of Powerwall ... and still got the same problem.**



P.S. Note this is a "secondary" machine - my primary machine continues to work just fine with 2.9.12 ... so it's not a Tesla issue ... and LOL I just hit a record high output of 6.62 kW from a 7.2kW spec'ed system.

**Reply by jasonacox** [reply]
Hi @hulkster  - I'm so glad you are doing this on your test machine! 

> File "/usr/local/lib/python3.10/site-packages/pypowerwall/cloud/pypowerwall_cloud.py", line 181, in connect
f"({sites[self.siteindex]['site_name']}) for {self.email}")
KeyError: 'site_name'

So, this is a problem. The setup is not completing because it is not getting a `site_name` back from the cloud. I believe some (all?) Solar Only systems may have this problem. I patched the code to handle this condition and have create a beta container for us to try if you are willing to help: jasonacox/pypowerwall:0.8.5t55-beta1

Step 1 - Edit the powerwall.yml file and find this section and update the image to match this:

```yml
    pypowerwall:
        image: jasonacox/pypowerwall:0.8.5t55-beta1
        container_name: pypowerwall
        hostname: pypowerwall
        restart: unless-stopped
```

Step 2 - Restart and then re-run setup:

```bash
# Restart stack with new version
./compose-dash.sh up -d

# Test to make sure your email is right
grep -E "^PW_EMAIL=.+" pypowerwall.env | cut -d= -f2

# Run setup
docker exec -it pypowerwall python3 -m pypowerwall setup -email=$(grep -E "^PW_EMAIL=.+" pypowerwall.env | cut -d= -f2)

# Restart pypowerwall
docker restart pypowerwall

# Verify
./verify.sh
```

If that works for you, I will push an update so upgrade.sh will work going forward.

**Reply by jasonacox** [reply]
I posted the wrong image initially - should be `jasonacox/pypowerwall:0.8.5t55-beta1` (edited)

**Reply by hulkster** [reply]
After modifying pypowerwall.yml to pull the beta1 version, everything now seems to be working - thanks a BUNCH @jasonacox 

Let me know when you push the update to Powerwall and I'll test this from scratch.

And if you want any type of further testing on a Solar Only install, just holler.

**Reply by jasonacox** [reply]
Awesome! Thank you! I'll get this updated and let you know. 🙏

**Reply by jasonacox** [reply]
Updated - You should be able to run`./upgrade.sh` to pull the latest.  Let me know if it works. Thanks for your help!

**Reply by hulkster** [reply]
Work great - really appreciate how easy upgrade.sh makes things - thank YOU for your excellent/responsive work! 

BTW, I noticed that powerwall.yml is specifying Grafana 9.1.2 ... but per #371, I'm wondering if it might make sense to push to a newer version(?)

**Reply by jasonacox** [reply]
That's great!  Thanks @hulkster 

I don't have a lot of time to do the testing rigor on a grafana upgrade. I'm working on some other priorities for the project right now. If you want to try it out, please do. Update the version in powerwall.yml and `./compose-dash.sh up -d`.  As always, good idea to back up everything, but reverting should be easy.; If you try it, let me know how it goes.

**Reply by hulkster** [reply]
Thanks @jasonacox for working those "other priorities" as I'm sure some good stuff coming!    ;-)

I commented over in #371 that I'll look at newer versions of Grafana and report back there with my findings.

I believe the "Solar Only Upgrade" thread has run it's course (kinda surprised that bug hadn't gotten discovered before) ... so I will try to remember to check back in a week and if no other comments, close it out.

**Reply by jasonacox** [reply]
I agree. However, it does seems like there can be different behaviors for different installs. I'm definitely going to enlist your help to keep an eye on solar-only system validation. 🙏 😉

**Comment by hulkster** [comment]
Tiny bit more info after spending a lot of time beating my head against the wall.

First, I confirmed that TZ is setup ... so it's not the issue discussed in #464 

I changed "PW_DEBUG" in pypowerwall.env to "YES" ... don't see any anything helpful.

Here's the output from running _"docker logs --details pypowerwall"_
05/14/2024 11:15:21 PM [proxy] [INFO] pyPowerwall [0.8.4] Proxy Server [t55] - HTTP Port 8675 - DEBUG
05/14/2024 11:15:21 PM [pypowerwall] [DEBUG] pypowerwall [0.8.4]
05/14/2024 11:15:21 PM [proxy] [INFO] pyPowerwall Proxy Started
05/14/2024 11:15:21 PM [pypowerwall.cloud.pypowerwall_cloud] [DEBUG]  -- cloud: Using site None for MYEMAIL
05/14/2024 11:15:21 PM [pypowerwall.cloud.pypowerwall_cloud] [DEBUG] Tesla cloud mode enabled
05/14/2024 11:15:22 PM [proxy] [ERROR] 'site_name'
05/14/2024 11:15:22 PM [proxy] [ERROR] Fatal Error: Unable to connect. Please fix config and restart.

I do "keep things local" and compose.env has this:
INFLUXDB_PORTS="127.0.0.1:8086:8086"
GRAFANA_PORTS="127.0.0.1:9000:9000"
PYPOWERWALL_PORTS="127.0.0.1:8675:8675"
WEATHER411_PORTS="127.0.0.1:8676:8676"

I was a little bit involved in the push to allow 127.0.0.1 (which works fine for version 2 on my primary machine) and in the latest release (4.3.1) as seen in the verify output, all the other services appear to be working correctly. 

Is it perhaps possible that with version 4 and the new cloud access via pypowerwall, that has some sort of bug with that?

**Comment by hulkster** [comment]
Closing out this thread - I appreciate how easy it is to upgrade to newer versions.

**Reply by jasonacox** [reply]
Thanks @hulkster! I appreciate the help and please don't hesitate to open up a new issue/discussion if you see anything wrong.
